src/rgw/rgw_role.h

```diff
@@ -0,0 +1,104 @@
+#ifndef CEPH_RGW_ROLE_H
+#define CEPH_RGW_ROLE_H
+
+class RGWRole
+{
+  static const string role_name_oid_prefix;
+  static const string role_oid_prefix;
+  static const string role_path_oid_prefix;
+
+  CephContext *cct;
+  RGWRados *store;
+  string id;
+  string name;
+  string path;
+  string creation_date;
+  string trust_policy;
+  map<string, string> perm_policy_map;
+
+  int store_info(bool exclusive);
+  int store_name(bool exclusive);
+  int store_path(bool exclusive);
+  int read_id(const string& role_name, string& role_id);
+  int read_name();
+  int read_info();
+  void set_id(const string& id) { this->id = id; }
+
+public:
+  RGWRole(CephContext *cct,
+          RGWRados *store,
+          string name,
+          string path,
+          string trust_policy)
+  : cct(cct),
+    store(store),
+    name(std::move(name)),
+    path(std::move(path)),
+    trust_policy(std::move(trust_policy)) {
+    if (this->path.empty())
+      this->path = "/";
+  }
+
+  RGWRole(CephContext *cct,
+          RGWRados *store,
+          string name)
+  : cct(cct),
+    store(store),
+    name(std::move(name)) {}
+
+  RGWRole(CephContext *cct,
+          RGWRados *store)
+  : cct(cct),
+    store(store) {}
+
+  RGWRole() {}
+
+  ~RGWRole() = default;
+
+  void encode(bufferlist& bl) const {
+    ENCODE_START(1, 1, bl);
+    ::encode(id, bl);
+    ::encode(name, bl);
+    ::encode(path, bl);
+    ::encode(creation_date, bl);
+    ::encode(trust_policy, bl);
+    ::encode(perm_policy_map, bl);
```

Comment:
All right. It seems untidy and pointless for something that hasn't been released but not that bad.